### ISSUE TITLE
[Testing 24.0] - tools/delete_imaging_upload.pl

### ISSUE THREAD

**Issue by cmadjar** [aces/Loris-MRI#680]
Blocked by #603

TODO:
- [x] Test deletion of an entire MRI upload (UploadID) with SQL and file backup (ensure the SQL and file backup repopulate the database correctly after the deletion)
- [x] Test deletion of an UploadID that has the same TarchiveID as an other UploadID - this should fail with an error message explaining there are multiple UploadID for a given TarchiveID
- [x] Test deletion of an UploadID with QC information attached - this should fail with proper error message
- [x] Test deletion of an UploadID with a parameter_form filled and the option `-form` and ensure the entries in the parameter form have been deleted
- [x] On a defaced dataset, test running the delete script with the `-defaced` option - this should delete only non-defaced MINC files for that UploadID
- [x] Run the delete script with option `-basename <FileBaseName>` specifying a basename for the MINC files to be deleted - only images matching that basename should be deleted for the UploadID

**Comment by zaliqarosli** [comment]
```
(loris-mri-python) lorisadmin@zrosli-dev:/data/loris/bin/mri$ delete_imaging_upload.pl -profile prod -backup_path /data/backups/imaging_upload_id89_backup -uploadID 89

Backing up files related to the upload(s) to delete...
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_t1_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_t2_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapDWI_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_002.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_dwi25_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fMRI_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fMRI_002.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapDWI_002.mnc

....

/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_003.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_004.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_005.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_006.mnc
/data/loris/data/tarchive/2018/DCM_2018-04-20_ImagingUpload-14-42-AEcK_S.tar

Added imaging_upload_restore.sql to the backup file.
Wrote /data/backups/imaging_upload_id89_backup.tar.gz
Successfully deleted data for upload 89.
```

extracting the tar.gz file results in the directories and files being re-created in the current directory i.e. relative path data/loris/data ... instead of /data/loris/data ... is this expected?
```
(loris-mri-python) lorisadmin@zrosli-dev:/data/backups$ tar -xzvf imaging_upload_id89_backup.tar.gz 
tar: Removing leading `/' from member names
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_t1_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_t2_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapDWI_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_001.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_fieldmapBOLD_002.mnc
/data/loris/data/assembly/947103/V3/mri/native/demo_947103_V3_dwi25_001.mnc

....

(loris-mri-python) lorisadmin@zrosli-dev:/data/loris/data/pic/947103$ cd /data/loris/data/assembly/947103/V3/mri/native/
(loris-mri-python) lorisadmin@zrosli-dev:/data/loris/data/assembly/947103/V3/mri/native$ ls -la
total 8
drwxrwx--- 2 lorisadmin www-data 4096 Oct 18 14:48 .
drwxrwx--- 4 lorisadmin www-data 4096 Oct  1 23:22 ..

...

(loris-mri-python) lorisadmin@zrosli-dev:/data/backups/data/loris/data/assembly/947103/V3/mri/native$ ls -la
total 643816
drwxrwxr-x 2 lorisadmin lorisadmin      4096 Oct 18 16:01 .
drwxrwxr-x 4 lorisadmin lorisadmin      4096 Oct 18 16:01 ..
-rwxrwx--- 1 lorisadmin lorisadmin       101 Oct  1 23:22 demo_947103_V3_dwi25_001.bval
-rwxrwx--- 1 lorisadmin lorisadmin       926 Oct  1 23:22 demo_947103_V3_dwi25_001.bvec
-rwxrwx--- 1 lorisadmin lorisadmin  21385379 Oct  1 23:22 demo_947103_V3_dwi25_001.mnc
-rwxrwx--- 1 lorisadmin lorisadmin  71885152 Oct  1 23:22 demo_947103_V3_dwi25_001.nii
-rwxrwx--- 1 lorisadmin lorisadmin       321 Oct  1 23:22 demo_947103_V3_dwi65_001.bval
```

SQL patch sources well though!

**Comment by zaliqarosli** [comment]
When testing item 4 "Test deletion of an UploadID with a parameter_form filled and the option -form and ensure the entries in the parameter form have been deleted", i got the following result:

```
(loris-mri-python) lorisadmin@zrosli-dev:/opt/loris/bin/mri$ delete_imaging_upload.pl -profile prod -backup_path /data/backups/uploadID_130_backup_imagingupload -uploadID 130 -form -ignore
Warning! The following files exist in the database but were not found on the file system:
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
	/data/loris/data/tarchive/DCM_2018-04-20_ImagingUpload-14-32-wJLsnI.tar
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_2_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_2_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
	/data/loris/data/pic/587630/loris_587630_V1_t2_002_445_check.jpg
	/data/loris/data/pic/587630/loris_587630_V1_dwi25_002_446_check.jpg
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.mnc
	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_t2_002.mnc


Backing up files related to the upload(s) to delete...
/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bval
/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bvec

DBD::mysql::db do failed: Cannot delete or update a parent row: a foreign key constraint fails (`zrosli_sandbox`.`bids_export_files`, CONSTRAINT `FK_bef_FileID` FOREIGN KEY (`FileID`) REFERENCES `files` (`FileID`)) at /opt/loris/bin/mri/tools/delete_imaging_upload.pl line 1865.
```

it doesn't look like this has anything to do with the -form flag being used, but it looks like if an upload has BIDS data, the delete script fails

**Comment by cmadjar** [comment]
> extracting the tar.gz file results in the directories and files being re-created in the current directory i.e. relative path data/loris/data ... instead of /data/loris/data ... is this expected?

I think this is expected. @nicolasbrossard can you confirm?

> SQL patch sources well though!

Cool!

**Comment by cmadjar** [comment]
> When testing item 4 "Test deletion of an UploadID with a parameter_form filled and the option -form and ensure the entries in the parameter form have been deleted", i got the following result:
> 
> ```
> (loris-mri-python) lorisadmin@zrosli-dev:/opt/loris/bin/mri$ delete_imaging_upload.pl -profile prod -backup_path /data/backups/uploadID_130_backup_imagingupload -uploadID 130 -form -ignore
> Warning! The following files exist in the database but were not found on the file system:
> 	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
> 	/data/loris/data/tarchive/DCM_2018-04-20_ImagingUpload-14-32-wJLsnI.tar
> 	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_2_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_2_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
> 	/data/loris/data/pic/587630/loris_587630_V1_t2_002_445_check.jpg
> 	/data/loris/data/pic/587630/loris_587630_V1_dwi25_002_446_check.jpg
> 	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
> 	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
> 	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.mnc
> 	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_t2_002.mnc
> 
> 
> Backing up files related to the upload(s) to delete...
> /data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bval
> /data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bvec
> 
> DBD::mysql::db do failed: Cannot delete or update a parent row: a foreign key constraint fails (`zrosli_sandbox`.`bids_export_files`, CONSTRAINT `FK_bef_FileID` FOREIGN KEY (`FileID`) REFERENCES `files` (`FileID`)) at /opt/loris/bin/mri/tools/delete_imaging_upload.pl line 1865.
> ```
> 
> it doesn't look like this has anything to do with the -form flag being used, but it looks like if an upload has BIDS data, the delete script fails

Indeed, the error reported above is unrelated to the -form flag of the script. I created a new issue for the delete script to be able to handle bids_export_files table: https://github.com/aces/Loris-MRI/issues/723

**Comment by zaliqarosli** [comment]
for the last case

- Run the delete script with option -basename <FileBaseName> specifying a basename for the MINC files to be deleted - only images matching that basename should be deleted for the UploadID:

```
(loris-mri-python) lorisadmin@zrosli-dev:/data/backups$ delete_imaging_upload.pl -profile prod -backup_path /data/backups/backup_delete_uploadID_80_basename_defaced -uploadID 80 -basename 'demo_676061_V5_t1_001.mnc' -ignore -defaced
Option -defaced cannot be used in conjunction with option -file. Aborting.
```

@cmadjar so looks like you can't use both basename and defaced options. is this fine? my only review is for `Option -defaced cannot be used in conjunction with option -file. Aborting.` to be changed to `Option -defaced cannot be used in conjunction with option -_basename_. Aborting.`

**Comment by zaliqarosli** [comment]
reran the -form option case:

```
(loris-mri-python) lorisadmin@zrosli-dev:/data/backups$ delete_imaging_upload.pl -profile prod -backup_path /data/backups/uploadID_130_backup_imagingupload -uploadID 130 -form -ignore
Warning! The following files exist in the database but were not found on the file system:
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_2_mri.mnc
	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_t2_002.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
	/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.mnc
	/data/loris/data/pic/587630/loris_587630_V1_dwi25_002_446_check.jpg
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_7d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_6d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-10-njrYmD/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
	/data/loris/data/pic/587630/loris_587630_V1_t2_002_445_check.jpg
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_8d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_5d1_mri.mnc
	/data/loris/data/trashbin/TarLoad-15-2-MKRbWy/dcc090_587630_v1_20180420_102155_2_mri.mnc
	/data/loris/data/tarchive/DCM_2018-04-20_ImagingUpload-14-32-wJLsnI.tar


Backing up files related to the upload(s) to delete...
/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bval
/data/loris/data/assembly/587630/V1/mri/native/loris_587630_V1_dwi25_002.bvec

mysqldump: Couldn't execute 'SELECT /*!40001 SQL_NO_CACHE */ * FROM `mri_parameter_form` WHERE CommentID IN (DDE_587630DCC0901053141524238782,587630DCC0901053141524238782)': Unknown column 'DDE_587630DCC0901053141524238782' in 'where clause' (1054)
Cannot run command mysqldump --no-create-info --compact --single-transaction --skip-extended-insert --no-tablespaces --where='CommentID IN (DDE_587630DCC0901053141524238782,587630DCC0901053141524238782)' --result-file=sql_backup_Ac9i  -h ace\-db\-2 -p13211_\!Password -u zrosli_admin zrosli_sandbox mri_parameter_form 2>&1 | fgrep -v 'mysqldump: [Warning] Using a password on the command line interface can be insecure.'. Aborting
```
